python/ray/dag/compiled_dag_node.py

```diff
@@ -438,30 +455,50 @@ def _compute(self, class_handle) -> bool:
             # Propagate it and skip the actual task. We don't need to wrap the
             # exception in a RayTaskError here because it has already been wrapped
             # by the previous task.
-            self.set_intermediate_buffer(exc)
+            self.set_intermediate_buffer(operation, exc)
             return False
 
+        channel_results = []
+        for entry in input_data:
+            if isinstance(entry, GPUFuture):
+                channel_result = entry.wait()
+            else:
+                channel_result = entry
+            channel_results.append(channel_result)
+
         resolved_inputs = []
         for task_input in self.task_inputs:
-            resolved_inputs.append(task_input.resolve(input_data))
+            resolved_inputs.append(task_input.resolve(channel_results))
 
         try:
             output_val = method(*resolved_inputs, **self.resolved_kwargs)
         except Exception as exc:
             output_val = _wrap_exception(exc)
-        self.set_intermediate_buffer(output_val)
+
+        if overlap_gpu_communication:
+            import cupy as cp
+
+            exec_event = cp.cuda.Event()
+            exec_event.record(cp.cuda.get_current_stream())
```

Comment:
Would be better to put this code in the constructor of `_GPUFuture`, so we don't leak GPU execution details into this function.

Ideally, I would like to see a generic Future constructor interface like:
```python
DAGNodeFuture(output_val: Any, cuda_stream: Optional[cp.cuda.Stream])
```

so we can use it for both GPU and non-GPU objects.

Reply:
Done.